extensions/cornerstone/src/services/ViewportService/CornerstoneViewportService.ts

```diff
@@ -548,115 +568,153 @@ class CornerstoneViewportService extends PubSubService
-            segmentation.id
-          );
-        }
+    // this only returns hydrated segmentations
+    const segmentations = segmentationService.getSegmentations();
+
+    for (const segmentation of segmentations) {
+      const toolGroupSegmentationRepresentations =
+        segmentationService.getSegmentationRepresentationsForToolGroup(
+          toolGroup.id
+        ) || [];
+
+      // if there is already a segmentation representation for this segmentation
+      // for this toolGroup, don't bother at all
+      const isSegmentationInToolGroup = toolGroupSegmentationRepresentations.find(
+        representation => representation.segmentationId === segmentation.id
+      );
+
+      if (isSegmentationInToolGroup) {
+        continue;
       }
-    }
 
-    const viewportInfo = this.getViewportInfo(viewport.id);
+      // otherwise, check if the hydrated segmentations are in the same FOR
+      // as the primary displaySet, if so add the representation (since it was not there)
+      const { id: segDisplaySetInstanceUID, type } = segmentation;
+      const segFrameOfReferenceUID = this._getFrameOfReferenceUID(
+        segDisplaySetInstanceUID
+      );
 
-    if (!viewportInfo) {
-      console.warn('Viewport info not defined for', viewport.id);
-    }
+      let shouldDisplaySeg = false;
 
-    const toolGroup = toolGroupService.getToolGroupForViewport(viewport.id);
-    csToolsUtils.segmentation.triggerSegmentationRender(toolGroup.id);
+      for (const displaySetInstanceUID of displaySetInstanceUIDs) {
+        const primaryFrameOfReferenceUID = this._getFrameOfReferenceUID(
+          displaySetInstanceUID
+        );
 
-    const initialImageOptions = viewportInfo.getInitialImageOptions();
+        if (segFrameOfReferenceUID === primaryFrameOfReferenceUID) {
+          shouldDisplaySeg = true;
+          break;
+        }
+      }
 
-    if (
-      initialImageOptions &&
-      (initialImageOptions.preset !== undefined ||
-        initialImageOptions.index !== undefined)
-    ) {
-      const { index, preset } = initialImageOptions;
+      if (!shouldDisplaySeg) {
+        return;
+      }
 
-      const { numberOfSlices } = csUtils.getImageSliceDataForVolumeViewport(
-        viewport
+      segmentationService.addSegmentationRepresentationToToolGroup(
+        toolGroup.id,
+        segmentation.id,
+        false, // already hydrated,
+        segmentation.type
       );
+    }
+  }
 
-      const imageIndex = this._getInitialImageIndex(
-        numberOfSlices,
-        index,
-        preset
-      );
+  private _addSegmentationRepresentationForSEGDisplaySet(
+    segDisplaySet: any,
+    viewport: any
+  ) {
+    const {
+      segmentationService,
+      toolGroupService,
+    } = this.servicesManager.services;
 
-      csToolsUtils.jumpToSlice(viewport.element, {
-        imageIndex,
-      });
+    const { referencedVolumeId } = segDisplaySet;
+    const referencedVolume = cache.getVolume(referencedVolumeId);
```

Comment:
Is there some way to fold these two lines into something based on the overlayDisplaySet so that it works for either seg or rt, or some future, as yet unknown overlay object?  I'd really like to not leak information from the segmentation service into this service, or rather leak things from the dicom-rt and dicom-seg modules here.  Those modules should be completely replaceable with new ones that do new types of things and still work.